### ISSUE TITLE
Add `list_deployed_models` to list models deployed on InferenceAPI

### ISSUE THREAD

**Issue by Wauplin** [huggingface/huggingface_hub#1557]
The new `InferenceClient` allows to run inference on the cloud. For each implemented task, new users that don't know which model to use can use the recommended model on the (free) Inference API (implemented in https://github.com/huggingface/huggingface_hub/pull/1510). Since the recommended model might not be suitable for all use cases, it is also possible to specify any model as input to the `InferenceClient`. While most models on the HuggingFace Hub is compatible with Inference API, waiting for the models to load can take a lot of time. Also, for large models the Inference API just cannot load them.

To ease user experience, it would be nice to have a utility `list_deployed_models` that would return a list of already deployed models for the user to try.

```py
class InferenceClient:
    ...
   
    def list_deployed_models(self, *, token: Optional[str] = None) -> Dict[str, List[str]]:
        ...
        return {"audio-to-audio": ["microsoft/speecht5_vc", ...], "text-generation": [...], ...}
```

This helper should help with discoverability.

---

Implementation-wise, this list of deployed models can be retrieved using the `"https://api-inference.huggingface.co/framework/{framework_name}"` where "framework_name" is a framework compatible with the Inference API. For example:
- https://api-inference.huggingface.co/framework/transformers
- https://api-inference.huggingface.co/framework/diffusers
- https://api-inference.huggingface.co/framework/text-generation-inference
- https://api-inference.huggingface.co/framework/sentence-transformers

Around ~30 frameworks are available but I think we should first focus on the main frameworks. Each url returns a list of dictionaries. Each item is a deployed model: `{"compute_type":"cpu","model_id":"microsoft/speecht5_vc","sha":"c418ba2144598f973d0fddc9fd5909a3af83de3d","task":"audio-to-audio"}`. What is interesting for us is `"model_id"` and `"task"`. What a `list_deployed_models` method would do is to loop through all main framework, loop  though all items and build an output dictionary with 'task' as key and a list of model ids as value.

Example:

```py
>>> from huggingface_hub import InferenceClient

>>> client = InferenceClient()
>>> client.deployed_models()["audio-to-audio"]
["microsoft/speecht5_vc", ...]
```

Happy to discuss the API design and update it if needed :) We still have to figure out what to do if several "sha" are deployed for a single model.

**Comment by Wauplin** [comment]
Current list of framework (can evolve). I doubt that all of them are useful. Maybe we should only focus on the top5 to avoid making X requests before returning.

```py
FRAMEWORKS = [
    "text-generation-inference",
    "transformers",
    "allennlp",
    "flair",
    "espnet",
    "asteroid",
    "speechbrain",
    "timm",
    "sentence-transformers",
    "spacy",
    "sklearn",
    "keras",
    "generic",
    "stanza",
    "adapter-transformers",
    "fasttext",
    "fairseq",
    "pyannote-audio",
    "doctr",
    "nemo",
    "fastai",
    "k2",
    "diffusers",
    "paddlenlp",
    "mindspore",
    "open_clip",
    "span-marker",
    "bertopic",
    "peft",
]
```

**Comment by julien-c** [comment]
(note that there's also an API to know if a given model is deployed or not. The `/status` one)

**Comment by osanseviero** [comment]
From inference API side of things, I'm not sure the underlying framework is that interesting. In other words, if a model does named-entity-recognition, I don't think people care that much if it's spacy or transformers if they want to know deployed models. I think people will be more interested in seeing all the models deployed for a certain task, but I'm also not sure if there's really that high demand for it

**Comment by Wauplin** [comment]
> I think people will be more interested in seeing all the models deployed for a certain task

Yes, that's exactly what I wanted to suggest in this issue!
I made the code snippet more explicit on what I was thinking about in term of return value. The underlying framework API that I suggested to use is just a way to get this information since the server doesn't have a list per task but a list per framework. Agree with you that the framework is not an important information.

> but I'm also not sure if there's really that high demand for it

Actually I would say yes. Especially for task that requires a lot of compute (e.g. models powered by TGI but also maybe some text-to-image models) users can't spin up the model themselves. Having an endpoint to list the available models gives an idea of which models have being manually curated by us, in addition to the models automatically running.

> note that there's also an API to know if a given model is deployed or not

Good to know. I've created a separate issue: https://github.com/huggingface/huggingface_hub/issues/1558
